File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/scope/EnvironmentHandle.java
package in.org.quicko.silo.client.scope;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonNode;
import in.org.quicko.silo.client.RequestOptions;
import in.org.quicko.silo.client.collections.CollectionCatalog;
import in.org.quicko.silo.client.collections.CollectionDefinition;
import in.org.quicko.silo.client.collections.CollectionHandle;
import in.org.quicko.silo.client.search.Search;
import in.org.quicko.silo.client.search.SearchPage;
import in.org.quicko.silo.client.search.SearchQuery;
import in.org.quicko.silo.client.search.SearchReach;
import in.org.quicko.silo.client.transport.ApiPath;
import in.org.quicko.silo.client.transport.TransportRequest;
import in.org.quicko.silo.client.variables.EnvironmentVariables;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * One environment, addressed by the name it was built with. Immutable, for the
 * reason {@link ProjectHandle} is.
 */
public final class EnvironmentHandle {
  private final ScopeReference scope;
  private final CollectionCatalog collections;
  private final EnvironmentVariables variables;

  public EnvironmentHandle(ScopeReference scope) {
    this.scope = scope;
    this.collections = new CollectionCatalog(scope);
    this.variables = new EnvironmentVariables(scope);
  }

  public String name() {
    return scope.environment();
  }

  public String project() {
    return scope.project();
  }

  /** The collections in this environment, and creating one. */
  public CollectionCatalog collections() {
    return collections;
  }

  /** This environment's variable values. */
  public EnvironmentVariables variables() {
    return variables;
  }

  /** An untyped collection, whose entries carry their fields as a map. */
  public CollectionHandle<Map<String, Object>> collection(String name) {
    return collection(name, new TypeReference<Map<String, Object>>() { });
  }

  /** A collection typed to the class its entries deserialise into. */
  public <F> CollectionHandle<F> collection(String name, Class<F> fieldsType) {
    return new CollectionHandle<>(scope, name, scope.transport().codec().typeOf(fieldsType));
  }

  /** A collection whose field type is generic, named by a {@link TypeReference}. */
  public <F> CollectionHandle<F> collection(String name, TypeReference<F> fieldsType) {
    JavaType resolved = scope.transport().codec().typeOf(fieldsType);
    return new CollectionHandle<>(scope, name, resolved);
  }

  public List<CollectionDefinition> schemas() {
    return schemas(RequestOptions.none());
  }

  /** Every schema in the scope, in one request. */
  public List<CollectionDefinition> schemas(RequestOptions options) {
    JsonNode body = scope.transport().json(
        TransportRequest.get(ApiPath.schemas(scope.project(), scope.environment()))
            .options(options)
            .build());

    List<CollectionDefinition> definitions = new ArrayList<>();
    body.path("items").forEach(item -> definitions.add(CollectionDefinition.fromWire(item)));
    return List.copyOf(definitions);
  }

  public SearchPage search(SearchQuery query) {
    return search(query, RequestOptions.none());
  }

  /** Searches every collection in this environment that the key can read. */
  public SearchPage search(SearchQuery query, RequestOptions options) {
    return new Search(
            scope.transport(), SearchReach.environment(scope.project(), scope.environment()))
        .run(query, options);
  }

  public RenameReport rename(String newName) {
    return rename(newName, RenameOptions.none());
  }

  public RenameReport rename(String newName, RenameOptions options) {
    return RenameReport.fromWire(scope.transport().json(
        TransportRequest.patch(ApiPath.environment(scope.project(), scope.environment()))
            .query("dry_run", options.dryRun() ? true : null)
            .query("expected_id", options.expectedId())
            .body(Map.of("name", newName))
            .options(options.request())
            .build()));
  }

  public void delete() {
    delete(DeleteOptions.none());
  }

  public void delete(DeleteOptions options) {
    scope.transport().empty(
        TransportRequest.delete(ApiPath.environment(scope.project(), scope.environment()))
            .query("force", options.force() ? true : null)
            .options(options.request())
            .build());
  }
}

File: packages/silo-client-dart/lib/src/scope/environment_handle.dart
import '../collections/collection_definition.dart';
import '../collections/collection_handle.dart';
import '../collections/collections.dart';
import '../collections/fields_converter.dart';
import '../request_options.dart';
import '../search/search.dart';
import '../search/search_page.dart';
import '../search/search_query.dart';
import '../search/search_reach.dart';
import '../transport/api_path.dart';
import '../transport/json_values.dart';
import '../transport/transport_request.dart';
import '../variables/environment_variables.dart';
import 'delete_options.dart';
import 'rename_options.dart';
import 'rename_report.dart';
import 'scope_reference.dart';

/// One environment by name. Immutable, like `ProjectHandle`.
final class EnvironmentHandle {
  EnvironmentHandle(this._scope) : collections = Collections(_scope), variables = EnvironmentVariables(_scope);

  final ScopeReference _scope;
  final Collections collections;
  final EnvironmentVariables variables;

  String get name => _scope.environment;

  String get project => _scope.project;

  /// Rows as maps. `withConverter` types them.
  CollectionHandle<Map<String, Object?>> collection(String name) => CollectionHandle(_scope, name, FieldsConverter.map);

  /// Every schema in the environment, in one request.
  Future<List<CollectionDefinition>> schemas([RequestOptions options = const RequestOptions()]) async {
    final body = await _scope.transport.json(
      TransportRequest('GET', ApiPath.schemas(_scope.project, _scope.environment), options: options),
    );
    return JsonValues.objects(body['items']).map(CollectionDefinition.fromJson).toList();
  }

  Future<SearchPage> search(SearchQuery query, [RequestOptions options = const RequestOptions()]) =>
      Search(_scope.transport, SearchReach.environment(_scope.project, _scope.environment)).run(query, options);

  Future<RenameReport> rename(String name, [RenameOptions options = const RenameOptions()]) async =>
      RenameReport.fromJson(
        await _scope.transport.json(
          TransportRequest(
            'PATCH',
            ApiPath.environment(_scope.project, _scope.environment),
            query: {if (options.dryRun) 'dry_run': true, 'expected_id': ?options.expectedId},
            body: {'name': name},
            options: options,
          ),
        ),
      );

  Future<void> delete([DeleteOptions options = const DeleteOptions()]) => _scope.transport.empty(
    TransportRequest(
      'DELETE',
      ApiPath.environment(_scope.project, _scope.environment),
      query: {if (options.force) 'force': true},
      options: options,
    ),
  );
}
